internal/pkg/upload/finalize.go

```diff
@@ -0,0 +1,145 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package upload
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"sort"
+	"strings"
+
+	"github.com/rs/zerolog/log"
+)
+
+func (u *Uploader) Complete(ctx context.Context, id string, transitHash string) (Info, error) {
+	// make sure document is freshly fetched, not cached
+	// so accurate status checking happens
+	info, err := FetchUploadInfo(ctx, u.bulker, id)
+	if err != nil {
+		return info, err
+	}
+
+	/*
+		Verify Upload
+	*/
+
+	// if already done, failed or deleted, exit
+	if !info.StatusCanUpload() {
+		return info, err
+	}
+
+	chunks, err := GetChunkInfos(ctx, u.bulker, info.DocID)
+	if err != nil {
+		return info, err
+	}
+	if !u.allChunksPresent(info, chunks) {
+		return info, ErrMissingChunks
+	}
+	if !u.verifyChunkInfo(info, chunks, transitHash) {
+		if err := SetStatus(ctx, u.bulker, info, StatusFail); err != nil {
+			log.Warn().Err(err).Str("fileID", info.DocID).Str("uploadID", info.ID).Msg("file upload failed chunk validation, but encountered an error setting the upload status to failure")
+		}
+		if err := DeleteChunksByQuery(ctx, u.bulker, info.Source, info.DocID); err != nil {
+			log.Warn().Err(err).Str("fileID", info.DocID).Str("uploadID", info.ID).Msg("file upload failed chunk validation, but encountered an error deleting left-behind chunk data")
+		}
+		return info, errors.New("file contents did not pass validation")
+	}
+
+	/*
+		Upload OK. Update status and save valid transithash
+	*/
+	if err := MarkComplete(ctx, u.bulker, info, transitHash); err != nil {
+		return info, err
+
+	}
+
+	return info, nil
+}
+
+func (u *Uploader) allChunksPresent(info Info, chunks []ChunkInfo) bool {
+	// check overall count
+	if len(chunks) != info.Count {
+		log.Warn().Int("expectedCount", info.Count).Int("received", len(chunks)).Interface("chunks", chunks).Msg("mismatch number of chunks")
+		return false
+	}
+
+	// now ensure all positions are accounted for, no gaps, etc
+	sort.Slice(chunks, func(i, j int) bool {
+		return chunks[i].Pos < chunks[j].Pos
+	})
+
+	for i, c := range chunks {
+		if c.Pos != i {
+			log.Warn().Int("expected", i).Interface("chunk", c).Msg("chunk position doesn't match. May be a gap in uploaded file")
+			return false
+		}
+	}
+
+	return true
+}
+
+func (u *Uploader) verifyChunkInfo(info Info, chunks []ChunkInfo, transitHash string) bool {
```

Comment:
nit: might return ```error``` instead of ```bool```, allows you to write better unit tests, testing different error cases, verifying the expected error